http-server/src/main/java/io/airlift/http/server/HttpServerChannelListener.java

```diff
@@ -0,0 +1,129 @@
+
+    public HttpServerChannelListener(DelimitedRequestLog logger)
+    {
+        this.logger = requireNonNull(logger, "logger is null");
+    }
+
+    @Override
+    public void onRequestBegin(Request request)
+    {
+        request.setAttribute(REQUEST_BEGIN_ATTRIBUTE, System.nanoTime());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onBeforeDispatch(Request request)
+    {
+        long requestBeginTime = (Long) request.getAttribute(REQUEST_BEGIN_ATTRIBUTE);
+        request.setAttribute(REQUEST_BEGIN_TO_DISPATCH_ATTRIBUTE, System.nanoTime() - requestBeginTime);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onRequestEnd(Request request)
+    {
+        long requestBeginTime = (Long) request.getAttribute(REQUEST_BEGIN_ATTRIBUTE);
+        request.setAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE, System.nanoTime() - requestBeginTime);
+    }
+
+    @Override
+    public void onResponseBegin(Request request)
+    {
+        request.setAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE, new ArrayList<Long>());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onResponseContent(Request request, ByteBuffer content)
+    {
+        List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
+        contentTimestamps.add(System.nanoTime());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onComplete(Request request)
+    {
+        List<Long> contentTimestamps = (List<Long>) request.getAttribute(RESPONSE_CONTENT_TIMESTAMPS_ATTRIBUTE);
+        long firstToLastContentTimeInMillis = -1;
+        if (contentTimestamps.size() > 0) {
+            firstToLastContentTimeInMillis = NANOSECONDS.toMillis(contentTimestamps.get(contentTimestamps.size() - 1) - contentTimestamps.get(0));
+        }
+        long beginToDispatchMillis = NANOSECONDS.toMillis((Long) request.getAttribute(REQUEST_BEGIN_TO_DISPATCH_ATTRIBUTE));
+        long beginToEndMillis = NANOSECONDS.toMillis((Long) request.getAttribute(REQUEST_BEGIN_TO_END_ATTRIBUTE));
+        logger.log(request,
+                request.getResponse(),
+                beginToDispatchMillis,
+                beginToEndMillis,
+                firstToLastContentTimeInMillis,
+                processContentTimestamps(contentTimestamps));
+    }
+
+    /**
+     * Calculate the summary statistics for the interarrival time of the onResponseContent callbacks.
+     */
+    @Nullable
+    private static DoubleSummaryStats processContentTimestamps(List<Long> contentTimestamps)
+    {
+        requireNonNull(contentTimestamps, "contentTimestamps is null");
+
+        // no content (HTTP 204)
+        if (contentTimestamps.size() == 0) {
+            return null;
+        }
+
+        DoubleSummaryStatistics statistics = new DoubleSummaryStatistics();
+
+        // If there was a single response chunk there won't be any interarrival times.
```

Comment:
Wouldn't it be more appropriate to have no inter-arrival stats in this case, then?

Reply:
Makes sense I think. Will update.